runtime/compiler/net/CommunicationStream.cpp

```diff
@@ -83,6 +83,60 @@ CommunicationStream::readMessage(Message &msg)
 #endif
    }
 
+void
+CommunicationStream::readMessageInFullCapacity(Message &msg)
+   {
+   msg.clearForRead();
+
+   // The message buffer storage and its capacity could be
+   // changed when the serialized size is set.
+   char *buffer = msg.getBufferStartForRead();
+   uint32_t bufferCapacity = msg.getBufferCapacity();
+
+   int32_t bytesRead = readOnceBlocking(buffer, bufferCapacity);
+
+   // bytesRead should be greater than 0 here, readOnceBlocking() throws
+   // an exception already if (bytesRead <= 0).
+   uint32_t serializedSize = (bytesRead >= sizeof(uint32_t)) ? ((uint32_t *)buffer)[0] : 0;
+   if (bytesRead < sizeof(uint32_t) ||
+       bytesRead > serializedSize) // Read more than the message size
+      {
+      throw JITServer::StreamFailure("JITServer I/O error: read error");
+      }
+
+   // serializedSize >= bytesRead
+   uint32_t bytesLeftToRead = serializedSize - bytesRead;
+
+   if (bytesLeftToRead == 0)
```

Comment:
`setSerializedSize` has the side effect of reallocating the buffer if needed, so the code could be simplified to
```
uint32_t bytesLeftToRead = serializedSize - bytesRead;
msg.setSerializedSize(serializedSize);
if (bytesLeftToRead > 0)
   readBlocking(msg.getBufferStartForRead() + bytesRead, bytesLeftToRead);
msg.deserialize();
```
In this case the message size is already written into the serialized version of the message, so maybe we need to split
`setSerializedSize()` into its components
1) setSerializedSize() just sets the size
2) expandBufferIfNeeded(size)  to expand the buffer.
If we do this then in the code here we only need `expandBufferIfNeeded()` which does the required copy anyway and in `readMessage()` we need to add `expandBufferIfNeeded()` just before `setSerializedSize()`

Reply:
I realized that `expandBufferIfNeeded` only copies up to what it thinks it's in the buffer, thus we may need to advance the cursor.

Reply:
> setSerializedSize() into its components

sounds good. 

> so the code could be simplified to
This might not work for the new read because `_curPtr` is not advanced when `expandBufferIfNeeded ` is called. `_curPtr`  still points to offset `0`. Whatever has been read into the older buffer will not get copied into the new buffer. As you mentioned the cursor needs to be advanced.